lib/client.ex

```diff
@@ -0,0 +1,153 @@
+defmodule Kaper.Client do
+
+  defmacro __using__(config) do
+    quote do
+      @conf unquote(config)
+      def conf, do: @conf
+      def list_tasks(opts \\ %{}) do
+        unquote(__MODULE__).list_tasks(conf(), opts)
+      end
+      def define_task(type, dbrps, script, status \\ :enabled, vars \\ %{}, id \\ "", template_id \\ "") do
+        unquote(__MODULE__).define_task(conf(), type, dbrps, script, status, vars, id, template_id)
+      end
+      def get_task(id, dot_view \\ "attributes", script_format \\ "formatted") do
+        unquote(__MODULE__).get_task(conf(), id)
+      end
+      def delete_task(id) do
+        unquote(__MODULE__).delete_task(conf(), id)
+      end
+      def enable_task(id) do
+        unquote(__MODULE__).enable_task(conf(), id)
+      end
+      def disable_task(id) do
+        unquote(__MODULE__).disable_task(conf(), id)
+      end
+      def define_template(type, script, id \\ "") do
+        unquote(__MODULE__).define_template(conf(), type, script, id)
+      end
+    end
+  end
+
+  def list_tasks(conf, opts \\ %{}) do
+    do_list_tasks(conf, opts)
+  end
+
+  defp do_list_tasks(conf, opts) do
+    attrs = remove_empty_values(%{
+      "pattern": Dict.get(opts, :pattern),
+      "fields": Dict.get(opts, :fields),
+      "dot-view": Dict.get(opts, :dotview),
+      "script-format": Dict.get(opts, :scriptformat),
+      "offset": Dict.get(opts, :offset),
+      "limit": Dict.get(opts, :limit),
+    })
+    query = URI.encode_query(attrs)
+    req_url = add_query(url("/kapacitor/v1/tasks", conf[:url]), query)
+    request :get, req_url, [], "", ""
+  end
+
+  def define_task(conf, type, dbrps, script, status \\ :enabled, vars \\ %{}, id \\ nil, template_id \\ nil) do
```

Comment:
This didn't work for me. I updated the `define_task` function to be private:

```elixir
  defp define_task(conf, type, dbrps, script, status \\ :enabled, vars \\ %{}, id \\ nil, template_id \\ nil) do
    do_define_task(conf, type, dbrps, script, status, vars, id, template_id)
  end

  defp do_define_task(conf, type, dbrps, script, status, vars, id , template_id) do
    attrs = %{
      "type": type,
      "dbrps": dbrps,
      "script": script,
      "status": status,
      "vars": vars,
      "id": id,
      "template_id": template_id,
    }
    req_url = url("/kapacitor/v1/tasks", conf[:url])

    body = Poison.encode!(attrs)
    request :post, req_url, [], "application/json", body
  end
```

I then modified the unit tests to use the `using` macro:

```elixir
  defmodule KapClient do
    use Kaper.Client, url: "http://0.0.0.0:9092"
  end

  test "define_task returns {:ok, response} if successful" do
    use_cassette "define_task" do
      {:ok, response} = KapClient.define_task "stream", [%{db: "kapacitor_example", rp: "default"}], "stream\n    |from()\n        .measurement('cpu')\n"

      assert response[:id] != ""
    end
  end
```

Running `mix test` produced the following output:

```shell
  1) test define_task returns {:error, reason} if internal server error (KaperTest)
     test/kaper_test.exs:49
     ** (UndefinedFunctionError) undefined function Kaper.Client.define_task/8
     stacktrace:
       (kaper) Kaper.Client.define_task([url: "http://0.0.0.0:9092"], "stream", [%{db: "kapacitor_example", rp: "default"}], "stream\n    |from()\n        .measurement('cpu')\n", :enabled, %{}, "", "")
       test/kaper_test.exs:52
```